### DISCUSSION TITLE
RFC: New Jotai v2 API

### DISCUSSION THREAD

**Discussion by dai-shi** [pmndrs/jotai#1514] [Ideas] [answered]
(This description will be updated over time. Please quote sentences when you add comments.)

## Background

Jotai is designed to fully support Concurrent React and Suspense for data fetching.
Despite of the simple api, the implementation has hacky things internally. If you ever see [src/core/store.ts](https://github.com/pmndrs/jotai/blob/01bf942d6f15eab4e17d46e5be316b4dfa3a17c5/src/core/store.ts), you know them.
Otherwise, it would have been a lot [simpler](https://github.com/pmndrs/jotai/blob/01bf942d6f15eab4e17d46e5be316b4dfa3a17c5/docs/guides/core-internals.mdx).

Other than the implementation complexity, there are some issues:
- useReducer behavior is confusing with derived atoms. We explain "extra re-renders with no commits" behavior is intentional, but not super comfortable, and we rely on `useMemo` for heavy computation.
- `unstable_createStore` #922 has very weird API for async values.
- Atoms are framework-agnostic, and most utils are too, but the library is tied to React.
- (maybe some more)

Furthermore, the biggest change recently is [the use RFC](https://github.com/reactjs/rfcs/pull/229). It indicates that throwing promises in a library is no longer supported (which is never supported), and needs `use`. However, it's not usable everywhere. Our current implementation will probably break.

## Why this change

When we released Jotai v1, we knew Suspense for data fetching is unstable. Our hope was, when it's finalized, we update our implementation with keeping Jotai v1 API.

It might be still possible, but as we see the use RFC, the general preference we should follow is async/await, not throwing promises behind the scene.

So, we propose new Jotai v2 API.

It will become more compatible with `use` and will technically get less benefit from Concurrent React.

## What will be breaking

After spending days and weeks, our conclusion is that this API is breaking change for async usage.

```js
// Current API
const asyncAtom = atom(async () => 'hello')
const derivedAtom = atom((get) => get(asyncAtom).toUppercase())

// New API
const asyncAtom = atom(async () => 'hello')
const derivedAtom = atom(async (get) => (await get(asyncAtom)).toUppercase())
```

This can be said a behavioral change, but the new type definition is different and  **incompatible** with the current one.

By the way, there's a big benefit with the current API. Not only it can skip async/await syntax, it can skip extra computation. It should perform well with Concurrent React too. But, for most cases, the new API should work well. We will also make `abortableAtom` first class in the new API, because otherwise there's no way to cancel computation after await. (There might still be room for improvements with caching, anyways.)

Even though the current `atom` and new `atom` works the same for sync atoms, for async atoms, the change is breaking. So, our plan is to provide a new entry point. And, hope users migrate to the new API gradually.

```js
// Current API
import { atom } from 'jotai'

// New API
import { atom } from 'jotai/vanilla'
```
(We might be able to expose the new API from `jotai` once all migration is done, and it may not be breaking for sync atoms.)

## Other major changes

- Split `/vanilla` and `/react` entry points for `jotai/utils`, `jotai/devtools` in addition to `jotai`
- Move other integrations out of this repo, like #1435 and #1473
- Stable `createStore` API
- Drop "scope" which has been confusing in usage and is unnecessary with `createStore` API
- Drop "versioned write" experiment as it won't be realized...

## Release plan

- **v1.11.0** pre-release v2 API (v1 API and v2 API coexist)
- **v1.12.0** deprecate integrations (jotai/query and so on)
- **v2.0.0** remove deprecated integrations and APIs, making all entry points with v2 API (minimal react version will be 17.0.0)

## Open problems

- How and when should we update the docs for this change?
  - 👉 #1515 has a migration guide.
- It may trigger more re-renders with async atoms. (Unconfirmed)
  - At least, it invokes more `read` functions, technically.
  - Maybe we could consider adding `unwrap` to simulate the current behavior?
  - 👉 #1515 has `unstable_unwrapAtom` for experimenting.

## Alternatives

- Drop the whole idea of "Async" API and just do others.
  - 👉 We'll move forward with "Async" API.
- Simply move to v2 with breaking changes without migration period.
  - 👉 We'll mostly do so.

**Comment by benrandja-akram** [comment]
Building for the future is always better since React is taking a big move here with "use" and async/await, so moving directly to v2 is a better option i guess while the changes are not that big.

**Reply by dai-shi** [reply]
Thanks for your comment. Yes, it's one good alternative. Let's see how the change will cause troubles during phase 1 and 2. We can then consider not deprecating some of them in phase 3.

**Reply by dai-shi** [reply]
I start to assume that the gradual migration is not very good in DX (especially with 3rd-party libraries), because two APIs are really incompatible with types (even without async atoms).

**Comment by dai-shi** [comment]
> [Phase 1] (v1.x) `unstable_` Async API

Hm, it might not be a good idea to have the prefix, because people have to modify it again. Probably, we should simply document it while it's "unstable".

**Comment by edongashi** [comment]
I wonder if it will be possible to `use(atom)` via the standard api. They mention that `use` will be applicable to multiple "wrapper" types, most obvious being promises.

**Reply by dai-shi** [reply]
My first trial: https://twitter.com/dai_shi/status/1569802927476457472
My current conclusion: https://twitter.com/dai_shi/status/1581856998085140480

**Reply by edongashi** [reply]
Interesting ideas. So `AsyncIterable` would represent a sequence of state transitions where we re-render at each step?

> to allow reading a cache from React context.

You mean `use(SomeContext)` or something else?

**Reply by dai-shi** [reply]
> AsyncIterable

yeah, I'd _expect_ so.

> use(SomeContext)

It's not allowed to use it internally. They require it to be used only in hooks.

**Comment by TKDev7** [comment]
Moving directly to v2 is a better option in my opinion as well.

**Reply by dai-shi** [reply]
Thanks for the comment. I still want to add the new api in v1, so that we have some period to migrate 3rd-party libs, but I get your point from the app developer perspective.

**Comment by Aslemammad** [comment]
> Atoms are framework-agnostic, and most utils are too, but the library is tied to React.

That's the problem with the current library, what we can do with the gradual API changes, is provide utils (in jotai-labs hopefully) for other component libraries, so we can gradually remove the notion of jotai being tied to React. So we would achieve the title of "Jotai: the framework-agnostic state-management library".
The recent trend around Signals can indeed inspire us in frameworks.

**Comment by dante01yoon** [comment]
Can you imagine when would be v2 release? 
I guess after use RFC closed or data fetching with suspense guide released?

**Reply by dai-shi** [reply]
> **v1.10.0** pre-release v2 API (v1 API and v2 API coexist)

I imagine to pre-release it in this year, hopefully soon.
#1515 is almost ready.
You can try it with codesandbox build.
https://ci.codesandbox.io/status/pmndrs/jotai/pr/1515
☝️ Find "Local Install Instructions"

> I guess after use RFC closed or data fetching with suspense guide released?

It's not tied to them. BUT, if they take down the RFC and take another route, we might need to re-consider. It's unlikely. We will go with Jotai v2 anyways, because there are many other benefits.

**Reply by dante01yoon** [reply]
thank you for quick respond. I'll look into it.

**Comment by 11bit** [comment]
One problem I found when experimenting with v2 api is that: 
1. I have to make any derived atom async if any dependency is async
2. it will suspend on every write even if async dependency is not used 

One example is an in-memory cache that loads something and then operates on it synchronously:
```js
const asyncAtom = atom(Promise.resolve("test"));

const derivedAtom = atom(async (get) => {
  return get(asyncAtom)
}, (get, set, update) => {
  // every write will suspend for a fraction of a second although asyncAtom doesn't have a promise anymore
  set(asyncAtom, update)
})
```

What is the correct way to make mixed atoms that may or may not suspend?

**Reply by dai-shi** [reply]
@11bit Thanks for the feedback!

If it's conditionally async, you want to avoid "async function" and use native promise instead.

In your simple case, it's easy.
```js
const asyncAtom = atom(Promise.resolve("test"));

const derivedAtom = atom((get) => {
  return get(asyncAtom)
}, (get, set, update) => {
  set(asyncAtom, update)
})
```

Let me know you have more complex examples.

**Reply by 11bit** [reply]
Thanks for the quick answer! What if have to use a value somehow (like filter through array fetched from somewhere)?

**Reply by dai-shi** [reply]
The `loadable` util should help to read async atom in sync.
We have also new `unstable_unwrap` util. This is still somewhat research field.

**Reply by 11bit** [reply]
By the way,  will the suspense behavior of `useAtom` change after react's `use` hook is released? As far as I understand currently there are some differences:
1. `use` doesn't suspend if a promise is resolved within a microtask
2. `useAtom` currently always suspend if the value is a promise

**Reply by dai-shi** [reply]
It should basically be the same:
https://github.com/pmndrs/jotai/blob/b3964f4bb6d27d5e34ea0b5741d831d3172a975f/src/react/useAtomValue.ts#L12-L41

**Comment by wobsoriano** [comment]
Thanks for this! Was able to experiment on a solid port with suspense working- https://github.com/wobsoriano/solid-jotai

**Reply by dai-shi** [reply]
Nice!

**Comment by csantos1113** [comment]
I'm giving v2 a try and I would love to know how can I initialize the Provider store with some data, within the react render cycle?

All examples I see in https://github.com/pmndrs/jotai/blob/main/tests/react/provider.test.tsx are setting the values outside of the rendering:

```
const store = createStore()
store.set(countAtom, 0)
store.set(petAtom, 'dog')

render(
  <Provider store={store}>
    <Display />
  </Provider>
)
```

But I'd love to do something like:

```
const uniqueStore = createStore();

const MyProvider = ({ count, children }: { count: number }) => {
  // but obviously this is bad
  uniqueStore.set(countAtom, count);
  return (
    <Provider store={uniqueStore}>
      {children}
    </Provider>
  );
}

render(
  <>
    <MyProvider count={2}><Display /></MyProvider>
    <MyProvider count={5}><Display /></MyProvider>
  </>
)
```

**Reply by dai-shi** [reply]
If you are migrating from v1, which has `initialValues` prop, then `useRef` is the equivalent solution.

```jsx
const MyProvider = ({ count, children }) => {
  const storeRef = useRef();
  if (!storeRef.current) {
    // lazy initialization of a ref is concurrent safe.
    storeRef.current = createStore();
    uniqueStore.set(countAtom, count);
  }
  return (
    <Provider store={storeRef.current}>
      {children}
    </Provider>
  );
}
```

**Reply by gsanta** [reply]
Hi the documentation suggests to use `Provider` with `initialValues` for setting up atoms for testing. Now to do this store setup for each test component seems to be quite overkill for me, and it would mean a lot of refactoring for bigger projects.

**Reply by dai-shi** [reply]
I updated the migration guide for a little easier pattern. How about creating a test provider?
```jsx
const TestHydrate = ({ initialValues, children }) => {
  useHydrateAtoms(initialValues)
  return children
}
const TestProvider = ({ initialValues, children }) => (
  <Provider>
    <TestHydrate initialValues={initialValues}>
      {children}
    </TestHydrate>
  </Provider>
)
```

**Reply by gsanta** [reply]
Thank you! This pattern is quite clear so I think it's a good replacement of the old behaviour!

**Reply by dai-shi** [reply]
#1744

**Comment by edongashi** [comment]
Hi @dai-shi. I have a question about the result of an atom writer:

> The write function can accept multiple arguments, and return a value.

In a discussion a few weeks ago you said that the setter is not guaranteed to run in the same frame.
Does this change mean that atom writers will always run immediately upon calling them?

**Reply by dai-shi** [reply]
Yes, v1 has experimental "versioned write", in which writers don't run immediately.

> Drop "versioned write" experiment as it won't be realized...

We dropped the experimental feature, so now it's guaranteed in v2.

**Reply by edongashi** [reply]
Thanks! I like this change, allows creating much more flexible "actions".

**Comment by dai-shi** [comment, marked as answer]
#1515 The first PR
#1671 Probably the last

**Comment by hackwaly** [comment]
Glad to see this. What i proposed here https://github.com/pmndrs/jotai/issues/1219#issuecomment-1150957541

**Reply by dai-shi** [reply]
Yeah, thanks for the discussion. My hesitation was that this *is* a big breaking change. Hope it goes well!

**Comment by katrinacodes** [comment]
I don't see it mentioned anywhere that `atomWithHash` was removed. Could you mention that in the migration guide and maybe provide a way to work around that?

**Reply by dai-shi** [reply]
Sorry for confusion.
We deprecated `atomWithHash` in `jotai/utils` in v1.10.0.
https://github.com/pmndrs/jotai/releases/tag/v1.10.0
Please use the one in `jotai-location`.
https://jotai.org/docs/integrations/location
